File: server/src/auth/password.ts
import crypto from 'node:crypto';

const DEFAULT_ITERATIONS = 30_000;
const KEY_LENGTH = 32;

export function hashPassword(password: string, iterations = DEFAULT_ITERATIONS): string {
  const salt = crypto.randomBytes(16);
  const dk = crypto.pbkdf2Sync(password, salt, iterations, KEY_LENGTH, 'sha256');
  return `pbkdf2$${iterations}$${salt.toString('base64')}$${dk.toString('base64')}`;
}

/** Strip spaces and compare student-id passwords case-insensitively. */
export function normalizeStudentPasswordPart(value: string): string {
  return value.replace(/\s+/g, '').toLowerCase();
}

/** Default student password is the student ID. */
export function verifyStudentLoginPassword(
  password: string,
  studentIdCode: string,
): boolean {
  const given = normalizeStudentPasswordPart(password);
  const id = normalizeStudentPasswordPart(studentIdCode);
  return Boolean(id) && given === id;
}

export function verifyPassword(password: string, stored: string): boolean {
  const parts = stored.split('$');
  if (parts.length !== 4 || parts[0] !== 'pbkdf2') return false;
  const iterations = Number.parseInt(parts[1], 10);
  if (!Number.isFinite(iterations)) return false;
  const salt = Buffer.from(parts[2], 'base64');
  const expected = Buffer.from(parts[3], 'base64');
  if (expected.length === 0) return false;
  const actual = crypto.pbkdf2Sync(password, salt, iterations, expected.length, 'sha256');
  if (actual.length !== expected.length) return false;
  return crypto.timingSafeEqual(actual, expected);
}

File: server/lib/src/auth/password.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

/// Salted PBKDF2-HMAC-SHA256 password hashing (no extra dependencies).
///
/// Stored format: `pbkdf2$<iterations>$<salt-b64>$<hash-b64>`
class PasswordHasher {
  static const _defaultIterations = 30000;
  static const _keyLength = 32;

  static String hash(String password, {int iterations = _defaultIterations}) {
    final salt = _randomBytes(16);
    final dk = _pbkdf2(utf8.encode(password), salt, iterations, _keyLength);
    return 'pbkdf2\$$iterations\$${base64Encode(salt)}\$${base64Encode(dk)}';
  }

  static bool verify(String password, String stored) {
    final parts = stored.split(r'$');
    if (parts.length != 4 || parts[0] != 'pbkdf2') return false;
    final iterations = int.tryParse(parts[1]);
    if (iterations == null) return false;
    final salt = base64Decode(parts[2]);
    final expected = base64Decode(parts[3]);
    final actual = _pbkdf2(
      utf8.encode(password),
      salt,
      iterations,
      expected.length,
    );
    return _constantTimeEquals(actual, expected);
  }

  static Uint8List _pbkdf2(
    List<int> password,
    List<int> salt,
    int iterations,
    int keyLength,
  ) {
    final hmac = Hmac(sha256, password);
    final hashLen = 32;
    final blocks = (keyLength / hashLen).ceil();
    final out = BytesBuilder(copy: false);

    for (var block = 1; block <= blocks; block++) {
      final blockBytes = Uint8List(4)
        ..buffer.asByteData().setUint32(0, block, Endian.big);
      var u = Uint8List.fromList(
        hmac.convert([...salt, ...blockBytes]).bytes,
      );
      final t = Uint8List.fromList(u);
      for (var i = 1; i < iterations; i++) {
        u = Uint8List.fromList(hmac.convert(u).bytes);
        for (var j = 0; j < t.length; j++) {
          t[j] ^= u[j];
        }
      }
      out.add(t);
    }
    return Uint8List.sublistView(out.toBytes(), 0, keyLength);
  }

  static Uint8List _randomBytes(int n) {
    final rng = Random.secure();
    return Uint8List.fromList(List.generate(n, (_) => rng.nextInt(256)));
  }

  static bool _constantTimeEquals(List<int> a, List<int> b) {
    if (a.length != b.length) return false;
    var diff = 0;
    for (var i = 0; i < a.length; i++) {
      diff |= a[i] ^ b[i];
    }
    return diff == 0;
  }
}
